File: 0290-word-pattern/0290-word-pattern.java
class Solution {
    public boolean wordPattern(String pattern, String s) {
        String[] words = s.trim().split("\\s+");
        if(pattern.length() != words.length){
            return false;
        }
        HashMap<Character, String> map = new HashMap<>();
        HashSet<String> set = new HashSet<>();

        for(int i = 0; i < pattern.length(); i++){
            char ch = pattern.charAt(i);
            String str = words[i];

            if(map.containsKey(ch)){
                if(!map.get(ch).equals(str)){
                    return false;
                }
            }else{
                if(set.contains(str)){
                    return false;
                }
                set.add(str);
                map.put(ch, str);
            }
        }
        return true;
    }
}

File: 0290-word-pattern/0290-word-pattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words = s.split()
        if len(words) != len(pattern):
            return False
        
        hashMap = {}
        for p, w in zip(pattern, words):
            key_p = ("p", p)
            key_w = ("w", w)

            if key_p in hashMap and hashMap[key_p] != w:
                return False
            if key_w in hashMap and hashMap[key_w] != p:
                return False
            hashMap[key_w] = p
            hashMap[key_p] = w
        return True